### ISSUE TITLE
MA0038: Do report diagnostic for benchmark methods

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#477]
BDN expects benchmark methods to be instance methods.
This of course means that **MA0038** is reported for any method that does not use any instance members.
Would it be possible not to report **MA0038** for methods on which a **BenchmarkDotNet.Attributes.BenchmarkAttribute** is applied?

**Comment by meziantou** [comment]
I'm considering disabling MA0038/MA0041 in favor of CA1822 as it is buggy and hard to implement correctly. However, I can implement a `DiagnosticSuppressor` to remove the diagnostic for some well-known libraries such as BenchmarkDotNet.

**Comment by drieseng** [comment]
I actually just switched from CA1822 to MA0038. Primary reason for doing so, is that you're a lot more responsive than MS.
I'm not sure I'm a fan of such a suppressor, unless if it can be configured to affect very specific patterns.

**Comment by meziantou** [comment]
A suppressor can choose what to suppress. You have access to the reported diagnostics, the location, the semantic model, etc. So, you can selectively disable specific patterns. This means you can detect methods reported by CA1822 having the attribute `Benchmark` and suppress them.

I think this is a good trade-off. CA1822 is more robust to report members that can be static, and Meziantou.Analyzer can suppress CA1822 on specific patterns.

I'll give it a try soon. I need to figure out how it actually works and how to test it as I've never implemented one.

**Comment by drieseng** [comment]
@meziantou, sounds great.

**Comment by drieseng** [comment]
@meziantou, problably shouldn't be asking questions at this hour, but:
* how do you enable / disable specific suppressions?
* are all suppressions enabled by default?

Can you disable a suppression by setting the severity of it's rule id to "none"?
For example:
```txt
dotnet_diagnostic.MAS0001.severity = none
```

If suppressions are not enabled by default, do you have to set the severity of the suprresion that you want to enable to a value that is higher or equal to the "active" severity level?
To always enable a given suppression, do you have to set its severity to error?
For example:
```txt
dotnet_diagnostic.MAS0001.severity = error
```

I haven't tried any of this yet, so this is probably no where near correct.

**Comment by meziantou** [comment]
Suppressions are pretty new to me, so I don't know all the details. They have an Id, so I think you should be able to disable them using the editorconfig. However, it doesn't work on my machine... That's being said, I think an analyzer should not suppress something where there is a doubt. There is no place for false-positives.

On my machine the suppressors are enabled by default. You can see that the method decorated with the attribute is not reported.

![image](https://user-images.githubusercontent.com/509220/227807125-f183679a-2782-444d-8736-bd816512d39d.png)

You can view suppressed diagnostic in the Error window in VS:

![image](https://user-images.githubusercontent.com/509220/227807143-05e8aa52-3b05-49fc-9852-361d47dc980d.png)

**Comment by drieseng** [comment]
I don't see a need to disable any of the suppressors that you implemented, but it may be useful to mention this somewhere.
